presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar.sql;
+
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.SqlInvokedScalarFunction;
+import com.facebook.presto.spi.function.SqlParameter;
+import com.facebook.presto.spi.function.SqlParameters;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+
+public class MapSqlFunctions
+{
+    private MapSqlFunctions() {}
+
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = false)
+    @Description("Returns the top N keys of the given map in descending order according to the natural ordering of its values.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint")})
+    @SqlType("array<K>")
+    public static String mapTopNKeys()
+    {
+        return "RETURN IF(n <= 0, fail('n must be a positive integer'), slice(reverse(array_sort(map_keys(input))), 1, n))";
+    }
+
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = true)
+    @Description("Returns the top N keys of the given map in sorted using the provided lambda comparator.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint"), @SqlParameter(name = "f", type = "function(K, K, int)")})
+    @SqlType("array<K>")
+    public static String mapTopNKeysComparator()
+    {
+        return "RETURN IF(n <= 0, fail('n must be a positive integer'), slice(reverse(array_sort(map_keys(input), f)), 1, n))";
```

Comment:
Actually, 0 is fine - should return empty array.

Reply:
Yes its fine but just thinking that what's the point of allowing 0 that means users don't want to return to anything which doesn't make sense. What do you think?

Reply:
Again - language semantics should be clean, precise and consistent. We should allow 0 like all others like SLICE

Reply:
Hmm..cool let me add it back quickly